### DISCUSSION TITLE
renovate not able update poetry.lock file

### DISCUSSION THREAD

**Discussion by vijaykumr** [renovatebot/renovate#31407] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted , latest docker image ( not full image)

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab

### Please tell us more about your question or problem

renovate fails to update the poetry.lock file . I had the set the variable GITHUB_COM_TOKEN with token having read permissions.
also used this in configuration for api.github.com
```javascript
 hostRules: [
    {matchHost: 'api.github.com', hostType: 'github', "readOnly": true, token:'{{ secrets.GITHUB_COM_TOKEN }}'} // token same as password?
  ],
```

I am renovate in aws eks cluster our organisation network layer . for outbound requests it has to go through nexus . so I had set this global.js
{
       "matchDatasources": ["github-releases"],
       "registryUrls": ["https://nexus.*****:8443/nexus/repository/python-containerbase/"]
    },

### Logs (if relevant)

debug logs
```
DEBUG: POST https://api.github.com/graphql = (code=ERR_INVALID_CHAR, statusCode=-1 retryCount=0, duration=-1) (repository=****, branch=renovate/all-python-patch)
DEBUG: Datasource connection error (repository=****)
       "datasource": "github-releases",
       "packageName": "containerbase/python-prebuild",
       "url": undefined,
       "errCode": "ERR_INVALID_CHAR"
 WARN: No tool releases found. (repository=****, branch=renovate/all-python-patch)
       "toolConfig": {
         "datasource": "github-releases",
         "packageName": "containerbase/python-prebuild",
         "versioning": "python"
       }
DEBUG: Failed to update poetry.lock file (repository=*****, branch=renovate/all-python-patch)
       "err": {
         "message": "No tool releases found.",
         "stack": "Error: No tool releases found.\n    at resolveConstraint (/usr/local/renovate/lib/util/exec/containerbase.ts:306:11)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at generateInstallCommands (/usr/local/renovate/lib/util/exec/containerbase.ts:357:27)\n    at prepareRawExec (/usr/local/renovate/lib/util/exec/index.ts:121:11)\n    at exec (/usr/local/renovate/lib/util/exec/index.ts:147:39)\n    at updateArtifacts (/usr/local/renovate/lib/modules/manager/poetry/artifacts.ts:210:5)\n    at managerUpdateArtifacts (/usr/local/renovate/lib/workers/repository/update/branch/get-updated.ts:505:12)\n    at getUpdatedPackageFiles (/usr/local/renovate/lib/workers/repository/update/branch/get-updated.ts:360:25)\n    at processBranch (/usr/local/renovate/lib/workers/repository/update/branch/index.ts:476:19)\n    at writeUpdates (/usr/local/renovate/lib/workers/repository/process/write.ts:166:17)\n    at update (/usr/local/renovate/lib/workers/repository/process/extract-update.ts:213:11)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:89:19)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:202:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:187:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
       }
```
log while reading package files
```
 "poetry": [
           {
             "deps": [
               {
                 "datasource": "github-releases",
                 "currentValue": ">= 3.9.7, < 3.9.20",
                 "managerData": {"nestedVersion": false},
                 "versioning": "pep440",
                 "depName": "python",
                 "depType": "dependencies",
                 "packageName": "containerbase/python-prebuild",
                 "commitMessageTopic": "Python",
                 "registryUrls": null,
                 "updates": [],
                 "skipReason": "ignored"
               },
```

**Comment by rarkins** [comment]
If you can't support an internet connection for runtime install then you can't use `binarySource=install` or the default image. Instead use the `-full` image which has package managers preinstalled and defaults to `binarySource=global`.

**Reply by vijaykumr** [reply]
I will a try @rarkins , using the full image

**Reply by vijaykumr** [reply]
No luck with that , I do see only this in logs . 
```
 "poetry": [
           {
             "deps": [
               {
                 "datasource": "github-releases",
                 "currentValue": ">= 3.9.7, < 3.9.20",
                 "managerData": {"nestedVersion": false},
                 "versioning": "pep440",
                 "depName": "python",
                 "depType": "dependencies",
                 "packageName": "containerbase/python-prebuild",
                 "commitMessageTopic": "Python",
                 "registryUrls": null,
                 "updates": [],
                 "skipReason": "ignored"
               },
```
Infact firewall is opened for datasources "github-releases" and in nexus the below Url has the remote set to https://github.com/containerbase/python-prebuild/. . I was expecting renovate should able to fetch the package from destination url as it has proxy set in nexus 

```
{
       "matchDatasources": ["github-releases"],
       "registryUrls": ["https://nexus.**:8443/nexus/repository/python-containerbase/"]
    },
```

**Reply by rarkins** [reply]
For whatever reason (most likely configuration error), that registryUrls is not being used.

You have also mixed up two different topics in one discussion here. The github-releases is never looked up as it has a skippedReason. The poetry error from before must relate to a different dependency

**Reply by vijaykumr** [reply]
so setting this datasource does not work as it was skipped . I have included this when it need internet to have the containerbase/python-prebuild installed as binarysource=install 
If binarysource=global , it doesnt make any sense to have this go through nexus and get package from github.

**Reply by vijaykumr** [reply]
Now I came to know what has happened from the logs it is understood, python version compatibility issue 
```
DEBUG: Failed to update poetry.lock file (repository=**** branch=renovate/all-python-patch)
       "err": {
         "cmd": "/bin/sh -c poetry update --lock --no-interaction boto3 pydantic pytest",
         "stderr": "The currently activated Python version 3.12.6 is not supported by the project (>= 3.9.7, < 3.9.20).\nTrying to find and use a compatible version. \n\nPoetry was unable to find a compatible version. If you have one, you can explicitly use it via the \"env use\" command.\n",
  ```

**Reply by rarkins** [reply]
Yes, that type of problem is unfortunately likely to occur if you do not allow Renovate to dynamically install the tools it needs using binarySource=install (which requires internet access).

**Reply by viceice** [reply]
you can try this url replacement to relocate dynamic containerbase installs

https://github.com/containerbase/base/blob/main/docs/custom-registries.md#custom-registries

you need to set manual constraints to renovate, so it skips looking up versions for tool install and instead passes the conteaint versions

**Reply by vijaykumr** [reply]
Interestingly in my case , some projects have the python version constraints as  **"currentValue": ">= 3.9.7, < 3.9.20",** and **python = "^3.12"** .
if I use the full image it did work for some of them and don't work few of them like it need python version in the range .
To work everyone if used the default image ( binarysource=install)
what should be the data sources it has to look ?
```
{
       "matchDatasources": ["github-releases"],
       "registryUrls": ["https://nexus.**:8443/nexus/repository/python-containerbase/"]
    },
 ```
 where https://nexus.**:8443/nexus/repository/python-containerbase/ is proxied for https://github.com/containerbase/python-prebuild/releases . but still does it work ? I doubt

**Reply by rarkins** [reply]
https://github.com/containerbase/base/blob/main/docs/custom-registries.md
